cEP-0002.md

```diff
@@ -0,0 +1,403 @@
+    * The content of the file, line-splitted
+    * It can also store other things about files as and when required.
+
+A prototype of the file proxy:
+
+```python
+from coala_utils.decorators import generate_eq
+
+
+@generate_eq("filename")
+class FileProxy:
+    def __init__(self, filename):
+        self.filename = filename
+        with open(self.filename) as filehandle:
+            self.content = filehandle.read()
+            self.lines = self.content.splitlines(True)
+
+    def __iter__(self):
+        return iter(self.lines)
+
+    def __hash__(self):
+        return hash(self.filename)
+```
+
+# Dependency Tracking
+
+A new class will be responsible for keeping track of the dependencies of different bear and it will work in the following manner:
+    * The list of all the bears is passed to the class.
+    * It takes this list and runs a method `check_circular_dependency` on it
+    to check if any bears in the list have circular dependencies. If they do
+    then we stop execution.
+    * Each bear will add it's dependencies to this tracker class.
+    * Whenever a bear completes it's execution the `resolve()` method of the
+    new class will be called with this bear as an argument.
+    * This would execute the marking of this bear as complete in the
+    dependency tracker.
+    * Once the marking is complete the tracker returns all bears which have no
+    dependencies remaining as pending, so that their tasks can be scheduled to
+    the process pool.
+
+Here is a prototype:
+
+```python
+from collections import defaultdict
+class CircularDependencyError(Exception):
+
+    @classmethod
+    def for_bears(cls, bears):
+        """
+        Creates the CircularDependencyError with a helpful message about the
+        dependency.
+        """
+        bear_names = [bear.name for bear in bears]
+
+        return cls("Circular dependency detected: " + " -> ".join(bear_names))
+
+
+class Dependencies:
+    def __init__(self):
+        self.dependency_dict = defaultdict(list)
+        self.dependency_set = set()
+
+    def _add_dependency(self, bear_instance, dependency):
+        self.dependency_dict[bear_instance].append(dependency)
+        self.dependency_set.add(dependency)
+
+    def add_bear_dependencies(self, bears):
+        """
+        Take the full bear list and check it for circular dependencies.Then
+        continue to add the dependenies of the bears to the class to be
+        resolved later.
+
+        :param bears: List of all bears.
+        """
+        self.check_circular_dependency(bears, [], [])
+        for bear_instance in bears:
+            for bear in bear_instance.BEAR_DEPS:
+                self._add_dependency(bear_instance, bear)
+
+    def check_circular_dependency(bears, resolved_bears, seen):
+        for bear in bears:
+            if bear in resolved_bears:
+                continue
+
+            missing = bear.missing_dependencies(resolved_bears)
+            if missing == []:
```

Comment:
if `missing_dependencies()` returns a list everytime, you can do:
`if not missing:`